### DISCUSSION TITLE
bin scripts disambiguation

### DISCUSSION THREAD

**Discussion by craigphicks** [npm/feedback#628] [CLI] [answered]
This post concerns bin scripts which will be called via `npx`, either explicitly or implicitly from a package.json `scripts` script.

- Problem:  
  - The `node_modules/.bin` namespace is prone to ambiguity as multiple packages might have executables with the same name.  I am aware that there some security issues have already been raised and resolved - however here I am more concerned with accidental collisions.  Example cases:
    - Case1: Supposing a monorepo, and tools developed as a local unpublished package with the monorepo not intended to be published to npm - name collisions and typo collisions are possible with past and future published executable sharing the same name. 
    - Case2: Different published packages accidentally share same executable name.

- Proposal: 
  - A new command `npx2` to accept packed-scoped bin executable names.  The interface could be either of these (or something else):
    - `npx2 [<scope-name>/]<package name>/<exe name>`
    - `npx2  [<scope-name>/]<package name> <exe name>`
  - Instead of looking in the local `node_modules/.bin` folder, it locates the executable from the `bin` property value(s) in `node_modules/[<scope-name>/]package name>/package.json`, and if it is not found, continues a hierarchical search the same as a `requires` search, with the following caveat:
  - Be able to specifiy a `package.json` property `"privateNames":[<entry>, ...]` where each entry is a string which is either a scope-name or a package name.  In case of an npx2 hierarchical search, the hierarchical search will go no higher than a package with a matching `"privateNames"` entry.  
  - Specify a `package.json` entry, e.g, `scriptsNpx2:true` that make the default scripts action `npx2` instead of `npx`. 
  - *Aside:: There are other scenarios where hierarchical searches could also be stopped in the same way by `"privateNames"` entries.  However these are each separate proposals that just happens to overlap with this issue, and their worthiness should not impact the worthiness of this posts proposal. They are not part of this posts proposal*
    - *The hierarchical search when adding a package with `npm -i`.*
    - *Hierarchical `requires` searches*  
  
 
To summarize:
1. Let `npx2` give executables the same namespacing as used by non-executable libs.  It makes sense as both executables and non-executables are distributed in npm style  packages.
2. Let `privateNames` properties define a hard namespace border for private executables inside a monorepo.  Monorepo's are very useful development paradigm, partly because they allow development of development tools (e.g., scripts) intended to be purely local inside the monorepo.  `privateNames` offers a "No means No" solution to prevent accidental external searches due to:  
  a. mismatching (or lacking) version numbers
  b. the executable is intended to be present but is not (e.g. clean without build) 
  c. typos where the scope is correct but the executable name has a typo.

**Comment by MylesBorins** [comment, marked as answer]
@craigphicks if you would like to make an explicit proposal for a new CLI feature the next step would be to make a proposal over at https://github.com/npm/rfcs

Thanks for the suggestion!

**Reply by craigphicks** [reply]
@MylesBorins - Oops. Reviewing the npx man page shows the proposal is already half completed.  (Of course!)  The only part that remains would be restricting execution (or installation) inside a local environment in case of scope or package name collisions (or typos).
